keras/utils/data_utils.py

```diff
@@ -561,18 +561,23 @@ def _wait_queue(self):
 
     def _run(self):
         """Submits request to the executor and queue the `Future` objects."""
-        sequence = list(range(len(self.sequence)))
-        self._send_sequence()  # Share the initial sequence
         while True:
+            # get new sequence indexes for each epoch
+            sequence = list(range(len(self.sequence)))
+            # Update pool with new sequence
+            self._send_sequence()
+
             if self.shuffle:
                 random.shuffle(sequence)
 
             with closing(self.executor_fn(_SHARED_SEQUENCES)) as executor:
-                for i in sequence:
+                for i, item_idx in enumerate(sequence):
                     if self.stop_signal.is_set():
                         return
-                    future = executor.apply_async(get_index, (self.uid, i))
-                    future.idx = i
+                    future = executor.apply_async(get_index, (self.uid, item_idx))
+                    future.idx = item_idx
+                    # last is True if item is last in current epoch
+                    future.last = i == len(sequence) - 1
```

Comment:
I tried to call `self.sequence.on_epoch_end` before `self._wait_queue()` call. This solves the issue of correct sequence length, but there are issues with the fact that sequence is shared between current thread and executor - now we can change state of the sequence before last future  from previous epochs has been done. This means we call `on_epoch_end` to start a new epoch, but executor still processes some of the `get_index` functions from previous epochs. 

That's why I added a flag to the future, which marks it as last in epoch. It will be true for last future in sequence for each epoch.  
This flag is used in `get` method to trigger `on_epoch_end` once executor not only excepted a task, but actually processed it.

Reply:
This adds a lot of complexity. If we wait for the queue to be empty, then there is no batch getting computed so it is safe to modify.

Reply:
That's why I modify sequence after last batch is consumed as well - to make sure no tasks are inside of executer when we call `self.sequence.on_epoch_end`. 

https://github.com/vikua/keras/blob/vikua/seq-gen-steps_per_epoch/keras/utils/data_utils.py#L606-L610
As you can use  `self.sequence.on_epoch_end` is called only when consumed future is last in epoch, result is computed (after `future.get(timeout=30)`) and before notifying `self.queue.task_done` to make sure new epoch is not started yet inside of `_run`.

Reply:
Sorry I don't think I get it.

When we call `_wait_queue()`, it will block till the last `task_done` is called. At this point, there is no batch getting processed, the queue is empty and no new future is getting queued. 


At this point, it should be safe to call `on_epoch_end`. If it's not the case, maybe there is a bug in `wait_queue`?

Reply:
I'm not sure 100% why it happens (trying to figure out), but if I move `self.sequence.on_epoch_end` after `_wait_queue()` - tests I added here will fail. For example in `test_fit_generator_dynamic_size_sequence_with_workers` second epoch will always have same length as first epoch, whereas it should be shorter.

Reply:
I think it is happening because `on_epoch_end` is called in the separate thread, whereas main thread is still spinning. 
So basically main thread consumes last items from the queue, updates queue with `task_done` an immediately recomputes steps_per_epoch and goes to next `queue.get` call. At this point `OrderedEnqueuer._run` thread might not called `self.sequence.on_epoch_end` yet. 
That's why new epoch gets wrong `steps_per_epoch`.

So in order to guarantee `steps_per_epoch` recomputed _after_ `self.sequence.on_epoch_end ` call I moved it to the main thread.

Reply:
@Dref360 I did same but using simpler way - just calling `sequence.on_epoch_end` in training loop 
commit is 8b7da25930fb5ac6826d834d7d01b26cbdd56471

This does same as previous version, but I think some tests in `data_utils_test.py` related to `OrderedEnqueuer` will fail, cause it doesn't call `self.sequence.on_epoch_end ` now

Reply:
What do you think - should be keep `self.sequence.on_epoch_end` call in `OrderedEnqueuer` (but have slightly more complicated version), or have `generator.on_epoch_end` call in training loop and change some tests in `data_utils_test.py `?

Reply:
@Dref360 I found another way to pull correct sequence length from `OrderedEnqueuer._run` thread into the main thread. I added another commit d8c1ab4ab86150a787256f693986c5a9378c43ae with the change.
Basically I use separate queue with just 1 slot to push information from run thread to main thread. This queue acts as an exchange between those thread. Whenever run thread is done thing an epoch it puts new length into this queue. Main thread on the other side pulls that queue for new sequence length. 
This helps to keep `OrderedEnqueuer` making `self.sequence.on_epoch_end` call. 
Also this helps to ensure correct order of `update sequence -> get length` events. 

So basically there are three ways right now 
- use `last` flag on the future (like I did initially) and call on_epoch_end in main thread in `OrderedEnqueuer.get`
- call on_epoch_end in main thread in training loop (simple, but some tests will need changing)
- use single slot queue to exchange information between main thread and `OrderedEnqueuer._run` thread.